m3/src/reporter_impl.cc

```diff
@@ -0,0 +1,438 @@
+      stream << minutes.count();
+      stream << "m";
+    }
+
+    stream << FormatDuration(nanos, 9);
+    stream << "s";
+  }
+
+  return stream.str();
+}
+
+std::string Reporter::Impl::BucketID(uint64_t bucket_id, uint64_t num_buckets) {
+  int len = std::to_string(bucket_id).size();
+
+  // Ensure the ID is at least 4 characters long.
+  len = std::max(len, 4);
+
+  std::string fmt = "%0" + std::to_string(len) + "d";
+
+  std::stringstream stream;
+  stream << boost::format(fmt) % bucket_id;
+
+  return stream.str();
+}
+
+std::string Reporter::Impl::FormatDuration(std::chrono::nanoseconds duration,
+                                           int precision) {
+  // Format the duration into the fraction of the given precision (i.e v /
+  // (10^p)). Omit any trailing zeros and the decimal when there is no
+  // fractional part.
+  double numerator = static_cast<double>(duration.count());
+  double divisor = std::pow(10, precision);
+  std::string str = std::to_string(numerator / divisor);
+
+  // Remove trailing zeros.
+  while (str[str.size() - 1] == '0') {
+    str.erase(str.size() - 1);
+  }
+
+  // Remove the decimal if it is the last character.
+  if (str[str.size() - 1] == '.') {
+    str.erase(str.size() - 1);
+  }
+
+  return str;
+}
+
+void Reporter::Impl::Run() {
+  std::shared_ptr<TCalcTransport> calc_transport(new TCalcTransport());
+
+  std::unique_lock<std::mutex> run_lock(run_mutex_);
+  while (run_) {
+    cv_.wait(run_lock);
+    if (!run_) {
+      // When the reporter is closed, this thread needs to drain the queue and
+      // flush any buffered metrics before it exits.
+      std::lock_guard<std::mutex> queue_lock(queue_mutex_);
+      while (queue_.size() > 0) {
+        auto metric = queue_.front();
+        Process(metric, calc_transport);
+        queue_.pop();
+      }
+      Flush();
+      return;
+    }
+
+    // While the queue is not empty this thread will continue to dequeue metrics
+    // without waiting on the condition variable.
+    while (true) {
+      std::unique_lock<std::mutex> queue_lock(queue_mutex_);
+      if (queue_.size() == 0) {
+        break;
+      }
+      auto metric = queue_.front();
+
+      // Release the queue lock so other threads can enqueue metrics while this
+      // thread processes its current batch of metrics.
+      queue_lock.unlock();
+
+      Process(metric, calc_transport);
+
+      queue_lock.lock();
+      queue_.pop();
+    }
+  }
+}
+
+void Reporter::Impl::Process(thrift::Metric metric,
+                             std::shared_ptr<TCalcTransport> calc_transport) {
+  TCompactProtocol calc_protocol(calc_transport);
+  metric.write(&calc_protocol);
+  auto size = calc_transport->size();
+
+  if (size > max_packet_size_) {
+    Flush();
+    calc_transport->flush();
+
+    thrift::MetricBatch batch;
+    batch.__set_commonTags(common_tags_);
+    batch.write(&calc_protocol);
+
+    // The current metric is not included in the batch of metrics that has just
```

Comment:
i don't follow - you've already done `metric.write(&calc_protocol)` before the size check, so why do you need to re-add it?

Reply:
nvm i see the calc_transport is just a glorified counter, so this works out.